File: src/main/java/org/dukcode/ps/codetree/trail02/chapter01/lesson03/IntroDetermineWhetherPalindromeIsPresent.java
package org.dukcode.ps.codetree.trail02.chapter01.lesson03;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.StringTokenizer;

/**
 * @see <a
 * href="https://www.codetree.ai/ko/trails/complete/curated-cards/intro-determine-whether-palindrome-is-present/description">LINK</a>
 */
public class IntroDetermineWhetherPalindromeIsPresent {

  private static BufferedReader br;
  private static BufferedWriter bw;
  private static StringTokenizer st;

  public static void main(String[] args) throws IOException {
    br = new BufferedReader(new InputStreamReader(System.in));
    bw = new BufferedWriter(new OutputStreamWriter(System.out));

    String word = br.readLine();

    bw.write(isPalindrome(word) ? "Yes" : "No");

    br.close();
    bw.close();

  }

  private static boolean isPalindrome(String word) {
    int n = word.length();
    for (int i = 0; i < n / 2; i++) {
      if (word.charAt(i) != word.charAt(n - 1 - i)) {
        return false;
      }
    }

    return true;
  }
}

File: src/main/kotlin/org/dukcode/ps/codetree/trail02/chapter01/lesson03/IntroDetermineWhetherPalindromeIsPresent.kt
package org.dukcode.ps.codetree.trail02.chapter01.lesson03

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter

private val br = BufferedReader(InputStreamReader(System.`in`))
private val bw = BufferedWriter(OutputStreamWriter(System.out))

/**
 * @see <a href="https://www.codetree.ai/ko/trails/complete/curated-cards/intro-determine-whether-palindrome-is-present/description">LINK</a>
 */
fun main() {
    val str = br.readLine()

    bw.write(if (str.isPalindrome()) "Yes" else "No")

    bw.flush()
}

private fun String.isPalindrome(): Boolean {
    var idx = 0
    while (idx < length / 2) {
        if (this[idx] != this[length - 1 - idx]) {
            return false
        }

        idx++
    }

    return true
}

private fun String.isPalindrome2(): Boolean {
    for (idx in 0..<(length / 2)) {
        if (this[idx] != this[length - 1 - idx]) {
            return false
        }
    }
    return true
}
